### PR TITLE
PSNR implementation

### LINKED ISSUE #2474: PSNR metric in metrics
## 🚀 Feature
Can we have [PSNR](https://en.wikipedia.org/wiki/Peak_signal-to-noise_ratio) metric in the metrics module?

### Motivation
This metric is used in Image generative models eg. super-resolution.

### PR DESCRIPTION
**Pull request by InCogNiTo124**
## What does this PR do?

Implements a new metric: PSNR (peak signal to noise ratio)

Fixes #2474 

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/metrics/regression.py [hunk 1]
```diff
@@ -187,3 +187,38 @@ def forward(self, pred: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
...
+    def forward(self, pred: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        mse = F.mse_loss(pred.view(-1), torch.view(-1))
+
+        # The calculation is troublesome because it is dependant of the maximum value possible.
+        # For integer inputs that should not be a problem (it's 255) but for floats there is a problem
+        # because the floats can be in [0, 1] range or they can be normalized with unknown mean and variance.
+        # Since mean and variance are unknown, we cannot know what's the maximum value to use in calculation.
+        # This implementation, therefore, finds the maximum empirically.
+        maximum = max(torch.max(torch.abs(pred)), torch.max(torch.abs(target)))
+        PSNR_base_e = 2*torch.log(maximum) - torch.log(mse)
```
Comment (justusschock) [inline pytorch_lightning/metrics/regression.py:222]: variable names should always be lowercase according to pep8

Comment (InCogNiTo124) [reply pytorch_lightning/metrics/regression.py]: Changed

**Comment by InCogNiTo124** [comment]
I've added tests and I validated them with `skimage.metrics.peak_signal_noise_ratio`, and it works (even the example in the comments)

**Comment by InCogNiTo124** [comment]
I've added tests and I validated them with `skimage.metrics.peak_signal_noise_ratio`, and it works (even the example in the comments)

**Comment by Borda** [comment]
> I've added tests and I validated them with `skimage.metrics.peak_signal_noise_ratio`, and it works (even the example in the comments)

add such validation test with `skimage` to our tests :rabbit:

**Comment by InCogNiTo124** [comment]
I wanted to, but I came to the conclusion that `skimage` is not in this project's requirements (I've installed the `devel` file) and I didn't want to change them :)

**Comment by awaelchli** [comment]
I agree with @Borda that checking with the skimage implementation for maybe random tensors would be the best. 
Could we also add a functional interface and coordinate with #2492 ?

**Comment by Borda** [comment]
it is fine to add `skimage` to `requirements/test.txt` :]

**Comment by rohitgr7** [comment]
You should add `PSNR` [in this file](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/metrics/__init__.py) too :)

**Comment by InCogNiTo124** [comment]
> I agree with @Borda that checking with the skimage implementation for maybe random tensors would be the best.

I'm having trouble figuring out how to test with random inputs, I don't know how do I pass randomly generated tensors from `pytest.param` into `skimage.metrics.peak_signal_noise_ratio`.


> Could we also add a functional interface and coordinate with #2492 ?

Sure :D

**Comment by rohitgr7** [comment]
Here also you should add PSNR :] https://github.com/PyTorchLightning/pytorch-lightning/blob/39a6435726e7a110df1189efe63094888e0c23be/pytorch_lightning/metrics/__init__.py#L49-L54

**Comment by InCogNiTo124** [comment]
> Here also you should add PSNR :]
> 
> https://github.com/PyTorchLightning/pytorch-lightning/blob/39a6435726e7a110df1189efe63094888e0c23be/pytorch_lightning/metrics/__init__.py#L49-L54

I totally missed it, it should be fixed now :)

File: pytorch_lightning/metrics/regression.py [hunk 2]
```diff
@@ -187,3 +187,38 @@ def forward(self, pred: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
...
+
+    def forward(self, pred: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        if self.data_range is None:
+            data_range = max(target.max() - target.min(), pred.max() - pred.min())
+        else:
+            data_range = torch.tensor(float(self.data_range))
+        mse = F.mse_loss(pred.view(-1), target.view(-1))
+        # numerical precision tricks
+        psnr_base_e = 2 * torch.log(data_range) - torch.log(mse)
+        return psnr_base_e * (10 / torch.log(self.base))  # change the logarithm basis
```
Comment (rohitgr7) [inline pytorch_lightning/metrics/regression.py:224]: You don't have to replicate the whole procedure again here. Just import from functional and use that one. :) Also add `reduction` parameter here.

Comment (InCogNiTo124) [reply pytorch_lightning/metrics/regression.py]: Refactored and added

BTW in case you didn't know, `pytest` says `'elementwise_mean'` is deprecated, it might be smart to refactor the default to `'mean'`. I left it that way to match rohitgr7's PR.

Comment (rohitgr7) [reply pytorch_lightning/metrics/regression.py]: Nice! Also, can you fix the pep8 issues mentioned above by @pep8speaks.

Comment (InCogNiTo124) [reply pytorch_lightning/metrics/regression.py]: @rohitgr7 I'm confused about the missing whitespace. I thought that was a bad thing :|

Comment (rohitgr7) [reply pytorch_lightning/metrics/regression.py]: hmm, don't know about that :| I never use an extra `,` after the last element in the list.

Comment (InCogNiTo124) [reply pytorch_lightning/metrics/regression.py]: Nvm I figured what was wrong

**Comment by InCogNiTo124** [comment]
I've got a few failing tests I don't know how to fix:
- I have no idea why `ufunc` is giving troubles in windows builds
- conda tests are missing skimage (weird because I've added in test requirements)
- there is a weird docker tag

**Comment by awaelchli** [comment]
> conda tests are missing skimage (weird because I've added in test requirements)

@InCogNiTo124 I'm pretty sure you need to add it also to environment.yml

**Comment by awaelchli** [comment]
> I have no idea why ufunc is giving troubles in windows builds


Lift numpy to 1.16.4 in requirements/base.txt, this was fixed in this version.

File: pytorch_lightning/metrics/functional/regression.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+    Example:
+
+        >>> from pytorch_lightning.metrics.regression import PSNR
+        >>> pred = torch.tensor([[0.0, 1.0], [2.0, 3.0]])
+        >>> target = torch.tensor([[3.0, 2.0], [1.0, 0.0]])
+        >>> metric = PSNR()
+        >>> metric(pred, target)
+        tensor(2.5527)
+        """
```
Comment (awaelchli) [inline pytorch_lightning/metrics/functional/regression.py:32]:
```suggestion
    """
```

Comment (InCogNiTo124) [reply pytorch_lightning/metrics/functional/regression.py]: Should be fixed now

**Comment by SkafteNicki** [comment]
If you could add a test similar to https://github.com/PyTorchLightning/pytorch-lightning/blob/a91b06ed1e5295efecdd7b51f4a3e6d95c829ecd/tests/metrics/functional/test_classification.py#L38-L72

to test on random tensors:
1. create two random tensors inside test
2. calculate value using skimage implementation and your implementation
3. assert results are same

**Comment by InCogNiTo124** [comment]
@SkafteNicki you mean alongside or instead of these tests I already wrote?

**Comment by SkafteNicki** [comment]
Just add a new test :)

**Comment by InCogNiTo124** [comment]
I googled the error in the TPU test, it seems that `PROJECT_ID` is not set.

**Comment by Borda** [comment]
> I googled the error in the TPU test, it seems that `PROJECT_ID` is not set.

Pls ignore TPU test for this moment, the issue is related to forked PRs... cc: @zcain117

**Comment by InCogNiTo124** [comment]
> Pls ignore TPU test for this moment, the issue is related to forked PRs... cc: @zcain117

In that case, this seems done then :)

**Comment by SkafteNicki** [comment]
Could you update CHANGELOG?
Also we need refer to this in the docs, but maybe we should just merge this and then fix that in #2492?

File: CHANGELOG.md [hunk 4]
```diff
@@ -8,7 +8,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 ## [unreleased] - YYYY-MM-DD
 
 ### Added
-
+- Added a new metric: peak signal-to-noise ratio ([#2483](https://github.com/PyTorchLightning/pytorch-lightning/pull/2483))
```
Comment (Borda) [inline CHANGELOG.md:11]:
```suggestion

- Added a PSNR metric: peak signal-to-noise ratio ([#2483](https://github.com/PyTorchLightning/pytorch-lightning/pull/2483))
```
